src/apps/wall/l7fw.lua

```diff
@@ -0,0 +1,353 @@
+function L7Fw:reject(pkt, len)
+   link.transmit(self.output.reject, self:make_reject_response())
+   self.rejected = self.rejected + 1
+
+   if self.logging == "on" then
+      self:log_packet("REJECT")
+   end
+
+   packet.free(self.current_packet)
+end
+
+-- called by pfmatch handler, forward packet
+function L7Fw:accept(pkt, len)
+   link.transmit(self.output.output, self.current_packet)
+   self.accepted = self.accepted + 1
+end
+
+function L7Fw:push()
+   local i       = assert(self.input.input, "input port not found")
+   local o       = assert(self.output.output, "output port not found")
+   local rules   = self.rules
+   local scanner = self.scanner
+
+   assert(self.output.reject, "output port for reject policy not found")
+
+   while not link.empty(i) do
+      local pkt  = link.receive(i)
+      local flow = scanner:get_flow(pkt)
+
+      -- so that pfmatch handler methods can access the original packet
+      self.current_packet = pkt
+
+      self.total = self.total + 1
+
+      if flow then
+         local name   = scanner:protocol_name(flow.protocol)
+         local policy = rules[name] or rules["default"]
+
+         self.current_protocol = name
+
+         if policy == "accept" then
+            self:accept(pkt.data, pkt.length)
+         elseif policy == "drop" then
+            self:drop(pkt.data, pkt.length)
+         elseif policy == "reject" then
+            self:reject(pkt.data, pkt.length)
+         -- handle a pfmatch string case
+         elseif type(policy) == "string" then
+            if self.handler_map[policy] then
+               -- we've already compiled a matcher for this policy
+               self.handler_map[policy](self, pkt.data, pkt.length, flow.packets)
+            else
+               local opts    = { extra_args = { "flow_count" } }
+               local handler = match.compile(policy, opts)
+               self.handler_map[policy] = handler
+               handler(self, pkt.data, pkt.length, flow.packets)
+            end
+         -- TODO: what should the default policy be if there is none specified?
+         else
+            self:accept(pkt.data, pkt.length)
+         end
+      else
+         -- TODO: we may wish to have a default policy for packets
+         --       without detected flows instead of just forwarding
+         self:accept(pkt.data, pkt.length)
+      end
+   end
+end
+
+function L7Fw:report()
+   local accepted, rejected, dropped =
+      self.accepted, self.rejected, self.dropped
+   local total = self.total
+   local a_pct = math.ceil((accepted / total) * 100)
+   local r_pct = math.ceil((rejected / total) * 100)
+   local d_pct = math.ceil((dropped / total) * 100)
+   print(("Accepted packets: %d (%d%%)"):format(accepted, a_pct))
+   print(("Rejected packets: %d (%d%%)"):format(rejected, r_pct))
+   print(("Dropped packets:  %d (%d%%)"):format(dropped, d_pct))
+end
+
+local logging_priority = bit.bor(LOG_USER, LOG_INFO)
```

Comment:
BitOps module is always accessible in LuaJIT but the docs recommend to explicitly require it:

> The suggested way to use the BitOp module is to add the following to the start of every
> Lua file that needs one of its functions:
> local bit = require("bit")
> This makes the dependency explicit, limits the scope to the current file and provides faster access to the bit.*

Source: http://bitop.luajit.org/api.html

This is more of a nit but I think it's helpful LuaJIT knowledge too.